akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -0,0 +1,401 @@
+    private case object InExecution extends WorkerState
+    private case object Idle extends WorkerState
+
+    val thread: Thread = tf.newThread(this)
+    @volatile private var workerState: WorkerState = NotStarted
+
+    def startWorker(): Unit = {
+      workerState = Idle
+      thread.start()
+    }
+
+    private def runCommand(command: Runnable) = {
+      workerState = InExecution
+      try
+        command.run()
+      finally
+        workerState = Idle
+    }
+
+    override def run(): Unit = {
+
+      /**
+       * Determines whether the worker can keep running or not.
+       * In order to continue polling for tasks three conditions
+       * need to be satisfied:
+       *
+       * 1) pool state is less than Shutting down or queue
+       * is not empty (e.g pool state is ShuttingDown but there are still messages to process)
+       *
+       * 2) the thread backing up this worker has not been interrupted
+       *
+       * 3) We are not in ShutDown state (in which we should not be processing any enqueued tasks)
+       */
+      def shouldKeepRunning =
+        (poolState < ShuttingDown || !q.isEmpty) &&
+          !Thread.interrupted() &&
+          poolState != ShutDown
+
+      var abruptTermination = true
+      try {
+        while (shouldKeepRunning) {
+          val c = q.poll()
+          if (c ne null) {
+            runCommand(c)
+            idleStrategy.reset()
+          } else // if not wait for a bit
+            idleStrategy.idle()
+        }
+        abruptTermination = false // if we have reached here, our termination is not due to an exception
+      } finally {
+        onWorkerExit(this, abruptTermination)
+      }
+    }
+
+    def stop() =
+      if (!thread.isInterrupted && workerState != NotStarted)
+        thread.interrupt()
+
+    def stopIfIdle() =
+      if (workerState == Idle)
+        stop()
+  }
+
+}
+
+object AffinityPool {
+
+  private def locked[T](l: Lock)(body: ⇒ T) = {
+    l.lock()
+    try {
+      body
+    } finally {
+      l.unlock()
+    }
+  }
+
+  private sealed trait PoolState extends Ordered[PoolState] {
+    def order: Int
+    override def compare(that: PoolState): Int = this.order compareTo that.order
+  }
+
+  // accepts new tasks and processes tasks that are enqueued
+  private case object Running extends PoolState {
+    override val order: Int = 0
+  }
+
+  // does not accept new tasks, processes tasks that are in the queue
+  private case object ShuttingDown extends PoolState {
+    override def order: Int = 1
+  }
+
+  // does not accept new tasks, does not process tasks in queue
+  private case object ShutDown extends PoolState {
+    override def order: Int = 2
+  }
+
+  // all threads have been stopped, does not process tasks and does not accept new ones
+  private case object Terminated extends PoolState {
+    override def order: Int = 3
+  }
+}
+
+final class BoundedTaskQueue(capacity: Int) extends AbstractBoundedNodeQueue[Runnable](capacity)
```

Comment:
@patriknw When you say internal here..do you mean that it should have the annotation or that it should be defined within the AffinityPool class . Or both ?